storage/mysql/mysql.go

```diff
@@ -159,3 +182,126 @@ func (s *Storage) ReadEntryBundle(ctx context.Context, index uint64) ([]byte, er
 	var entryBundle []byte
 	return entryBundle, row.Scan(&entryBundle)
 }
+
+// Add is the entrypoint for adding entries to a sequencing log.
+func (s *Storage) Add(ctx context.Context, entry *tessera.Entry) (uint64, error) {
+	// TODO: Return index if the value is already stored.
+
+	return s.queue.Add(ctx, entry)()
+}
+
+// sequenceBatch writes the entries from the provided batch into the entry bundle files of the log.
+//
+// This func starts filling entries bundles at the next available slot in the log, ensuring that the
+// sequenced entries are contiguous from the zeroth entry (i.e left-hand dense).
+// We try to minimise the number of partially complete entry bundles by writing entries in chunks rather
+// than one-by-one.
+//
+// TODO: Separate sequencing and integration for better performance.
+func (s *Storage) sequenceBatch(ctx context.Context, entries []*tessera.Entry) error {
+	// Return when there is no entry to sequence.
+	if len(entries) == 0 {
+		return nil
+	}
+
+	// Get a Tx for making transaction requests.
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	// Defer a rollback in case anything fails.
+	defer func() {
+		if err := tx.Rollback(); err != nil && err != sql.ErrTxDone {
+			klog.Errorf("Failed to rollback in sequenceBatch: %v", err)
+		}
+	}()
+
+	// Get tree size from checkpoint.
+	row := tx.QueryRowContext(ctx, selectCheckpointByIDSQL, checkpointID)
```

Comment:
I don't think this has an exclusive lock at this point? This may be repeating a race condition from the early experiment where a sufficiently high load of write traffic can cause multiple batches to be sequenced in parallel. Each of these performs this read, starts sequencing at the same checkpoint size, and then stomp on the other writes, with only the last winning.

The way this was fixed in the experiment was with a `SELECT ... FOR UPDATE`: https://github.com/AlCutter/betty/blob/mysql/storage/tsql/sql.go#L195